crates/ruff_python_semantic/src/analyze/class.rs

```diff
@@ -100,23 +97,47 @@ pub fn any_super_class(
 
 /// Return `true` if `class_def` is a class that has one or more enum classes in its mro
 pub fn is_enumeration(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
-    any_qualified_name(class_def, semantic, &|qualified_name| {
-        matches!(
-            qualified_name.segments(),
-            [
-                "enum",
-                "Enum" | "Flag" | "IntEnum" | "IntFlag" | "StrEnum" | "ReprEnum" | "CheckEnum"
-            ]
-        )
+    any_qualified_name(class_def, semantic, &|expr| {
+        semantic
+            .resolve_qualified_name(map_subscript(expr))
+            .is_some_and(|qualified_name| {
+                matches!(
+                    qualified_name.segments(),
+                    [
+                        "enum",
+                        "Enum"
+                            | "Flag"
+                            | "IntEnum"
+                            | "IntFlag"
+                            | "StrEnum"
+                            | "ReprEnum"
+                            | "CheckEnum"
+                    ]
+                )
+            })
     })
 }
 
 /// Returns `true` if the given class is a metaclass.
 pub fn is_metaclass(class_def: &ast::StmtClassDef, semantic: &SemanticModel) -> bool {
-    any_qualified_name(class_def, semantic, &|qualified_name| {
-        matches!(
-            qualified_name.segments(),
-            ["" | "builtins", "type"] | ["abc", "ABCMeta"] | ["enum", "EnumMeta" | "EnumType"]
-        )
+    any_qualified_name(class_def, semantic, &|expr| match expr {
+        Expr::Call(ast::ExprCall { func, .. }) => {
+            // Ex) `class Foo(type(Protocol)): ...`
+            semantic
+                .resolve_qualified_name(func.as_ref())
+                .is_some_and(|qualified_name| {
+                    matches!(qualified_name.segments(), ["" | "builtins", "type"])
+                })
+        }
+        _ => semantic
+            .resolve_qualified_name(map_subscript(expr))
+            .is_some_and(|qualified_name| {
+                matches!(
+                    qualified_name.segments(),
+                    ["" | "builtins", "type"]
+                        | ["abc", "ABCMeta"]
+                        | ["enum", "EnumMeta" | "EnumType"]
```

Comment:
Can any of these be subscripted? I don't think so...

Reply:
builtins.type can

```pycon
>>> class X(type[int]): pass
... 
>>> X.__mro__
(<class '__main__.X'>, <class 'type'>, <class 'object'>)
```